src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMProbeTest.java

```diff
@@ -25,78 +28,136 @@
-                        .withHeader("Content-Type", "application/json")
-                        .withBodyFile("body-yolo-contents-8rd37.json")));
+        githubApi.stubFor(get(urlEqualTo("/repos/cloudbeers/yolo/contents/subdir?ref=refs%2Fpull%2F1%2Fmerge")).willReturn(aResponse().withStatus(404)).atPriority(0));
+
+        createProbeForPR(1);
+
+        assertTrue(probe.stat("README.md").exists());
+        assertFalse(probe.stat("subdir").exists());
         assertFalse(probe.stat("subdir/Jenkinsfile").exists());
     }
 
     @Issue("JENKINS-54126")
     @Test
-    public void statWhenRootIs404AndCacheOnThenOff() throws Exception {
+    public void statWhenRoot404andThenIncorrectCached() throws Exception {
         GitHubSCMSource.setCacheSize(10);
-        githubApi.stubFor(get(urlPathEqualTo("/repos/cloudbeers/yolo/contents/")).withHeader("Cache-Control", containing("max-age")).willReturn(aResponse().withStatus(404)));
-        githubApi.stubFor(get(urlPathEqualTo("/repos/cloudbeers/yolo/contents/")).withHeader("Cache-Control", absent())
-                .willReturn(aResponse()
-                    .withStatus(200)
-                    .withHeader("Content-Type", "application/json")
-                    .withBodyFile("body-yolo-contents-8rd37.json")));
 
+        createProbeForPR(9);
+
+        // JENKINS-54126 happens when:
+        // 1. client asks for a resource "Z" that doesn't exist
+        // ---> client receives a 404 response from github and caches it.
+        // ---> Important: GitHub does not send ETag header for 404 responses.
+        // 2. Resource "Z" gets created on GitHub but some means.
+        // 3. client (eventually) asks for the resource "Z" again.
+        // ---> Since the the client has a cached response without ETag, it sends "If-Modified-Since" header
+        // ---> Resource has changed (it was created).
+        //
+        // ---> EXPECTED: GitHub should respond with 200 and data.
+        // ---> ACTUAL: GitHub server lies, responds with incorrect 304 response, telling client that the cached data is still valid.
+        // ---> THE BAD: Client cache believes GitHub - uses the previously cached 404 (and even adds the ETag).
+        // ---> THE UGLY: Client is now stuck with a bad cached 404, and can't get rid of it until the resource is _updated_ again or the cache is cleared manually.
+        //
+        // This is the cause of JENKINS-54126. This is a pervasive GitHub server problem.
+        // We see it mostly in this one scenario, but it will happen anywhere the server returns a 404.
+        // It cannot be reliably detected or mitigated at the level of this plugin.
+        //
+        // WORKAROUND (implemented in the github-api library):
+        // 4. the github-api library recognizes any 404 with ETag as invalid. Does not return it to the client.
+        // ---> The github-api library automatically retries the request with "no-cache" to force refresh with valid data.
```

Comment:
I love the explanation here, thanks for doing it.

Reply:
Also, IDK if you did this already, but it might be worth opening a support ticket with GitHub telling them the behavior we are seeing now that we know what is happening. We have a workaround so it doesn't matter if they fix it, but it's always nice to report bugs.

Reply:
Opened and they have not responded at all.